src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSTree.java

```diff
@@ -0,0 +1,322 @@
+            tree.put(parent, new PSTreeNode());
+        }
+        tree.get(parent).addChild(id);
+    }
+
+    /**
+     * Deletes nodes unreachable from the root and returns the set of deleted nodes.
+     */
+    public Set<String> removeUnreachableNodes() {
+        final Set<String> reachable = traverse();
+        final Set<String> unreachable = new HashSet<>(tree.keySet());
+        unreachable.removeAll(reachable);
+        retainNodes(reachable);
+        return unreachable;
+    }
+
+    /**
+     * Because of the piecemeal way we allow the tree to be constructed, we can end up with invalid tree structures.
+     * Check that tree structure contains valid pointers and is connected.
+     */
+    public void checkStructure() {
+
+        //Check child-parent pointers are consistent
+        for (final String id : tree.keySet()) {
+            final PSTreeNode n = tree.get(id);
+            for (final String child : n.getChildren()) {
+                if (!tree.containsKey(child)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has non-existent child " + child);
+                }
+                if (!tree.get(child).getParent().equals(id)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has child " + child + ", whose parent is " + tree.get(child).getParent() + " instead of " + id);
+                }
+            }
+            final String parent = n.getParent();
+            if (parent != null) {
+                if (!tree.containsKey(parent)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has non-existent parent " + parent);
+                }
+                if (!tree.get(parent).getChildren().contains(id)) {
+                    throw new UserException.BadInput("Malformed tree detected. Node " + id + " has parent " + parent + ", which does not have child " + id);
+                }
+            }
+        }
+
+        //Check disconnected sets of nodes using a traversal from the root
+        final Set<String> reachable = traverse();
+        if (reachable.size() != tree.size()) {
+            final Set<String> unreachable = tree.keySet();
+            unreachable.removeAll(reachable);
+            final String nodesList = getAbbreviatedNodeListString(unreachable);
+            throw new UserException.BadInput("Malformed tree detected. Tree is disconnected. " + unreachable.size() + " of " + tree.size() + " nodes were unreachable: " + nodesList);
+        }
+    }
+
+    /**
+     * Find all nodes reachable from the root
+     */
+    private Set<String> traverse() {
+        final Queue<String> queue = new LinkedList<>();
+        final Set<String> visited = new HashSet<>(tree.size());
+        queue.add(root);
+        while (!queue.isEmpty()) {
+            final String id = queue.poll();
+            if (!visited.contains(id)) { //checked visited in case there are cycles
```

Comment:
Yes, for some reason I thought there was a cycle check in checkStructure() but apparently not. Added the exception here, where it will now be thrown during checkStructure() as well.